core/src/main/java/pl/allegro/tech/embeddedelasticsearch/ElasticSearchInstaller.java

```diff
@@ -109,19 +113,58 @@ private File fileRelativeToInstallationDir(String... path) {
     }
 
     private Path download(URL source) throws IOException {
-        File target = new File(baseDirectory, constructLocalFileName(source));
+        File target = new File(getDownloadDirectory(), constructLocalFileName(source));
+        File statusFile = new File(target.getParentFile(), target.getName() + ELS_PACKAGE_STATUS_FILE_SUFFIX);
+
+        removeBrokenDownload(target, statusFile);
+
         if (!target.exists()) {
-            logger.info("Downloading : " + source + " to " + target + "...");
-            FileUtils.copyURLToFile(source, target);
-            logger.info("Download complete");
+            download(source, target, statusFile);
+        } else if (!statusFile.exists() && maybeDownloading(target)) {
+            waitForDownload(target, statusFile);
+        } else if (!statusFile.exists()) {
+            throw new IOException("Broken download. File '"  + target + "' exits but status '" + statusFile + "' file wash not created");
         } else {
             logger.info("Download skipped");
         }
         return target.toPath();
     }
 
+    private void removeBrokenDownload(final File target, final File statusFile) throws IOException {
+        if (target.exists() && !statusFile.exists() && !maybeDownloading(target)) {
+            logger.info("Removing broken download file {}", target);
+            FileUtils.forceDelete(target);
+        }
+    }
+
+    private boolean maybeDownloading(final File target) {
+        // Poor mans check based on assumption that if other process downloads e file should be modified
```

Comment:
This check is also used to check if some other thread or jvm is currently downloading installation package in such case current thread will wait for the other to finish download. Race situations are still possible but should still occur very rarely then without this check 

The other case where this is needed is when some kills build during download. In such cases on next run this check is used to conditionally remove stalled partial downloads and then thread will proceed new download.